piptools/build.py

```diff
@@ -0,0 +1,152 @@
+from __future__ import annotations
+
+import collections
+import contextlib
+import os
+import pathlib
+import tempfile
+from dataclasses import dataclass
+from importlib import metadata as importlib_metadata
+from typing import Iterator
+
+import build
+import build.env
+import pyproject_hooks
+from pip._internal.req import InstallRequirement
+from pip._internal.req.constructors import install_req_from_line, parse_req_from_line
+
+PYPROJECT_TOML = "pyproject.toml"
+
+
+@dataclass
+class ProjectMetadata:
+    extras: tuple[str, ...]
+    requirements: tuple[InstallRequirement, ...]
+    build_requirements: tuple[InstallRequirement, ...]
+
+
+def build_project_metadata(
+    src_file: str,
+    build_distributions: tuple[str, ...],
+    isolated: bool = True,
+) -> ProjectMetadata:
+    """
+    Return the metadata for a project.
+
+    Uses the ``prepare_metadata_for_build_wheel`` hook for the wheel metadata
+    if available, otherwise ``build_wheel``.
+
+    Uses the ``prepare_metadata_for_build_{dist}`` hook for each ``build_distributions``
+    if available.
+
+    :param src_file: Project source file
+    :param build_distributions: A tuple of build distributions to get the dependencies
+                                of (``sdist`` or ``wheel`` or ``editable``).
+    :param isolated: Whether to run invoke the backend in the current
+                     environment or to create an isolated one and invoke it
+                     there.
+    """
+
+    src_dir = os.path.dirname(os.path.abspath(src_file))
+    with _create_project_builder(src_dir, isolated) as builder:
+        metadata = _build_project_wheel_metadata(builder)
+        extras = tuple(metadata.get_all("Provides-Extra") or ())
+        requirements = tuple(
+            _prepare_requirements(metadata=metadata, src_file=src_file)
+        )
+        build_requirements = tuple(
+            _prepare_build_requirements(
+                builder=builder,
+                src_file=src_file,
+                distributions=build_distributions,
+                package_name=metadata.get_all("Name")[0],
+            )
+        )
+        return ProjectMetadata(
+            extras=extras,
+            requirements=requirements,
+            build_requirements=build_requirements,
+        )
+
+
+@contextlib.contextmanager
+def _create_project_builder(
+    src_dir: str, isolated: bool = False
+) -> Iterator[build.ProjectBuilder]:
+    builder = build.ProjectBuilder(
+        os.fspath(src_dir),
+        runner=pyproject_hooks.default_subprocess_runner,
+    )
+
+    if not isolated:
+        yield builder
```

Comment:
I can see that the `build` package does not install the build requirements in this case but I don't understand why. Do you happen to know the answer?

Reply:
`--build-isolation/--no-build-isolation` flags reflect to `isolated` param:

```
$ pip-compile --help
...
  --build-isolation / --no-build-isolation
                                  Enable isolation when building a modern
                                  source distribution. Build dependencies
                                  specified by PEP 518 must be already
                                  installed if build isolation is disabled.
...
```


This essentially means that build requirements should be provided by the user.